pkg/kv/kvserver/replica_consistency.go

```diff
@@ -500,9 +500,27 @@ func CalcReplicaDigest(
 	var timestampBuf []byte
 	hasher := sha512.New()
 
+	// Request quota from the limiter in chunks of quotaBatchSize, to amortize the
+	// overhead of the limiter when reading many small KVs.
+	var budget int64
+	const quotaBatchSize = int64(4 << 10) // 4 KiB
+	wait := func(size int64) error {
+		if size <= budget {
+			budget -= size
+			return nil
+		}
+		batches := (size - budget + quotaBatchSize - 1) / quotaBatchSize
+		tokens := batches * quotaBatchSize
+		if err := limiter.WaitN(ctx, tokens); err != nil {
+			return err
+		}
+		budget += tokens - size
```

Comment:
I'm not sure I completely understand these calculations. Can't we just allow bursts and do something straightforward like:

```go
var batchSize int64
const targetBatchSize = int64(4 << 10) // 4 KiB
wait := func(size int64) error {
  if batchSize += size; batchSize >= targetBatchSize {
    if err := limiter.WaitN(ctx, batchSize); err != nil {
      return err
    }
    batchSize = 0
  }
  return nil
}
```

Reply:
The calculation rounds up to multiples of `quotaBatchSize`. So, after the first row (if it's small) we will have acquired more than its size, and the next few entries will be free. This way the quota acquisition is almost always ahead of reading from the iterator.

You proposal is cleaner, but acquires quota post-factum, so in between acquisitions there is a bigger chance of a thundering herd between tasks.

Both solutions are "leaky" in some sense though, but the one in this PR is not leakier than the status quo: right now we acquire quota post-factum, but on every row; after the change we acquire it post-factum for a single row, and there is a chance that a few next rows are free.

I can mix the 2 ideas though. I think we don't need to always request a multiple of the batch size, so the maths can be simpler. But I think we should try to acquire upfront still.

Reply:
I don't think we need to overengineer this. The default setting here is 8 MB/s, which means that we're pausing about 30 ms for each batch (assuming only one check). I don't think reading 256 KB is expensive enough to matter. It's more about making sure it doesn't go full throttle.

Reply:
Done. Also, added a `WaitN` in the end, for any "uncommitted" quota.